### PR TITLE
add beforeExit callback to Resource api

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/543

I am not super happy with the fact that each implementation has to worry about this.... but I don't see a way around it :(.

### CODE REVIEW

File: build/lib/src/resource/resource.dart [hunk 1]
```diff
@@ -23,10 +24,16 @@ class Resource<T> {
   /// disposed.
   final DisposeInstance<T> _userDispose;
 
+  /// Optional method which is called before the process is going to exit. This
```
Comment (natebosch) [inline build/lib/src/resource/resource.dart]: [nit] blank line following first sentence. Similar in other doc comments.

File: build/lib/src/resource/resource.dart [hunk 2]
```diff
@@ -53,8 +60,16 @@ class Resource<T> {
 class ResourceManager {
   final _resources = new Set<Resource>();
 
+  /// The [Resource]s that we need to call `beforeExit` on. We have to hang on
+  /// to these forever, but they should be small in number, and we don't hold
+  /// on to the instances.
```
Comment (natebosch) [inline build/lib/src/resource/resource.dart]: "we don't hold on to the instances" - this isn't clear to me, could you rephrase to make it more clear we mean the instances of the created resource, not the `Resource` instance?

File: build/lib/src/resource/resource.dart (hunk 1 above)
Comment (jakemac53) [reply build/lib/src/resource/resource.dart]: done

File: build/lib/src/resource/resource.dart (hunk 2 above)
Comment (jakemac53) [reply build/lib/src/resource/resource.dart]: done
